### PR TITLE
extmod/modframebuf: added GS4_HMSB format

### PR DESCRIPTION
**Pull request by kamikaze**
Added 4bpp grayscale format. Tested on SSD1322 display

### CODE REVIEW

File: extmod/modframebuf.c [hunk 1]
```diff
@@ -115,7 +171,7 @@ static inline uint32_t getpixel(const mp_obj_framebuf_t *fb, int x, int y) {
 }
 
 STATIC void fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, int h, uint32_t col) {
-    if (x + w <= 0 || y + h <= 0 || y >= fb->height || x >= fb->width) {
```
Comment (kamikaze) [inline extmod/modframebuf.c]: I have a question about this part. What was a purpose of the part that I have changed? As for me it was pointless check but I could be wrong. So let's discuss this

Comment (pfalcon) [inline extmod/modframebuf.c]: > What was a purpose of the part that I have changed?

That's not a good question to ask ;-). (Akin to standing in a factory asking: what does that red button which I just pressed does?) Here it's clear that it checks whether a rectangle to be drawn and framebuffer's bounding rectangle intersect. If they don't, there's nothing to be drawn. If they do, there's something to draw, though perhaps a clipped part of the rectangle requested.

What's important is that this line has no relation to "adding GRAY4 format" whatsoever, so it shouldn't be a part of this change. And of course, before submitting a separate pull request to change it, you would first need to understand what it does (or otherwise just leave it alone). Take an example of https://github.com/micropython/micropython/pull/2751 - it's sad that @dpgeorge decided not to merge it, but it was done right - you addressed one issue in it, and it didn't contain unrelated changes. Please keep up doing it like that, not mix in different kind of changes like here.

Comment (kamikaze) [inline extmod/modframebuf.c]: asking is always a good thing ) and actually fill_rect is a related part to gray4 fill rect ) I will return previous check. But as I see it - there is a need to check for zero height/width too. No need to call subroutines if there is nothing to do

Comment (kamikaze) [inline extmod/modframebuf.c]: Extracted into separate PR #2771

File: extmod/modframebuf.c [hunk 2]
```diff
@@ -97,13 +97,69 @@ STATIC void rgb565_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, i
...
 
+// Functions for GRAY4 format
+
+STATIC void gray4_setpixel(const mp_obj_framebuf_t *fb, int x, int y, uint32_t col) {
+    uint8_t *pixel = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+
+    if (x % 2) {
+        *pixel = ((uint8_t)col & 0x0F) | (*pixel & 0xF0);
+    }
+    else {
```
Comment (dpgeorge) [inline extmod/modframebuf.c]: Please note style: all curly brackets on the one line.

File: extmod/modframebuf.c [hunk 3]
```diff
@@ -97,13 +97,69 @@ STATIC void rgb565_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, i
...
+    }
+
+    return ((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1] >> 4;
+}
+
+STATIC void gray4_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, int h, uint32_t col) {
+    uint8_t *pixel_pair = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+    uint8_t col_cleaned = (uint8_t)col & 0x0F;
+    uint8_t col_shifted_left = col_cleaned << 4;
+    uint8_t colored_pixel_pair = col_shifted_left | col_cleaned;
```
Comment (dpgeorge) [inline extmod/modframebuf.c]: I don't think you gain much by having these explicit variables.  Can you check if the code size reduces (or stays the same) if you just repeat these in the code below?  You can just do `col &= 0x0f` here and then use `col << 4` below.

**Review by dpgeorge — COMMENTED** [review]
This is really an addition of a generic 4 bit-per-pixel format, it's not tied to gray pixels only.  But considering we call the 16bpp format RGB565, I guess GRAY4 is an acceptable name.

But then also consider that there are 4 ways to have 4bpp (like with 1bpp): vert-msb, vert-lsb, horiz-msb, horiz-lsb.  I think you have horiz-msb here, so it should be named something like GRAY4_HMSB.

Also, please provide (in a separate commit but same PR) tests :)  You only need to provide unit tests for the new code you wrote: get pixel, set pixel, rect with various combinations of odd/even width and odd/even starting positions.

File: extmod/modframebuf.c (hunk 3 above)
Comment (kamikaze) [inline extmod/modframebuf.c]: First there was a for-loop instead of memset() and it had larger impact on speed. But it's still a good idea to calculate values (that will not change inside loops) outside. No unneeded memory allocations, no unneeded cpu usage aswell. Drawing graphics is not a strong part especially in case of I2C/SPI + bigger screen resolution. So I think it's better to do drawing as fast as possible everywhere it is possible. If we don't want a slideshow instead of animation.

And again - readability )

the only thing to optimize by removing is col_cleaned indeed.

**Comment by kamikaze** [comment]
about GRAY4. I just don't know how to name it. Actually in datasheets it is called "grayscale" or "GS". So I could call it as "GS4_HMSB" or "GRAY4_HMSB"

File: extmod/modframebuf.c (hunk 3 above)
Comment (kamikaze) [inline extmod/modframebuf.c]: renamed to GS4_HMSB (this name actually influences the fw size)

**Comment by kamikaze** [comment]
renamed GRAY4 to GS4_HMSB (this name actually influences the fw size)

**Comment by kamikaze** [comment]
believe me or not, but removing pre-calculated variables INCREASED the fw size ))

before:
```
   text    data     bss     dec     hex filename
 302876     352   28240  331468   50ecc build-PYBV11/firmware.elf
```

after:

```
   text    data     bss     dec     hex filename
 302888     352   28240  331480   50ed8 build-PYBV11/firmware.elf
```

**Comment by kamikaze** [comment]
Added tests

**Comment by kamikaze** [comment]
What else can I do to get this accepted? )

File: tests/extmod/framebuf4.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+fbuf.fill(0xf0)
+printbuf()
+fbuf.fill(0x0a)
+printbuf()
+fbuf.fill(0xfa)
+printbuf()
+fbuf.fill(0x0f)
+printbuf()
+fbuf.fill(0x00)
+printbuf()
```
Comment (dpgeorge) [inline tests/extmod/framebuf4.py]: I don't think all of these fill tests are needed... surely just fill(0x0f), fill(0xa0) would be enough?

File: tests/extmod/framebuf4.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+fbuf.fill_rect(1, 0, w//2, 1, 0x10)
+printbuf()
+fbuf.fill_rect(0, h-4, w//2+1, 4, 0xaf)
+printbuf()
+fbuf.fill_rect(0, h-4, w//2+1, 4, 0xb0)
+printbuf()
+fbuf.fill_rect(0, h-4, w//2, 4, 0xaf)
+printbuf()
+fbuf.fill_rect(0, h-4, w//2, 4, 0xb0)
+printbuf()
```
Comment (dpgeorge) [inline tests/extmod/framebuf4.py]: Can you remove a few of these fill_rect tests.  You should only need to check once that a colour with the 4 low bits cleared (and arbitrary bits set in the top 4 bits) does indeed just clear the rect.

Comment (kamikaze) [inline tests/extmod/framebuf4.py]: but what about odd/even start/end pixel and/or width? I should check for this

Comment (dpgeorge) [inline tests/extmod/framebuf4.py]: Yes, but you've got twice as many tests above as are needed.  Eg you don't really need both `fill_rect(0, h-4, w//2, 4, 0xaf)` and `fill_rect(0, h-4, w//2, 4, 0xb0)`, just the first one of those 2 would be enough.

Comment (kamikaze) [inline tests/extmod/framebuf4.py]: but if I will skip second "0xb0" fill - then buffer will not be clean (all zeros) so I can't be sure if the resulting state is correct and (and actually it will not be, because x,y,h,w are changing every time)

File: tests/extmod/framebuf4.py (hunk 4 above)
Comment (kamikaze) [inline tests/extmod/framebuf4.py]: removed other calls

**Comment by kamikaze** [comment]
reduced excessive tests

**Comment by dpgeorge** [comment]
Thank you!  It's been merged, see 406fb3cb6084e5b3c6489a99b30affd35330d3fd.
